### ISSUE TITLE
feature: permutation

### ISSUE THREAD

**Issue by thjashin** [tensorflow/tensorflow#5281]
This is a feature request on a Tensor creation function based on problems met when using tf.transpose.
Suppose I want to transpose a Tensor `a` (not variable) into a new Tensor, exchanging its two dimensions. Indices of the two dimensions are represented by two Tensors `b` and `c`.
From what I understand, there are only one complicated way of generating the argument `perm` in tf.transpose to achieve this in Tensorflow. That's by using tf.select twice.

``` python
dims = tf.range(5)
b = tf.constant(1)
c = tf.constant(3)
b_mask = tf.cast(tf.one_hot(b, 5), tf.bool)
c_mask = tf.cast(tf.one_hot(c, 5), tf.bool)
bs = tf.ones(5, tf.int32) * b
cs = tf.ones(5, tf.int32) * c
perm_1 = tf.select(b_mask, cs, dims)
perm_2 = tf.select(c_mask, bs, perm_1)
# ==> perm2: [0, 3, 2, 1, 4]
```

I personally feel this is too redundant for achieving such a simple operation. Could it be possible to add some op that generate permutations when it is not possible to be achieved by tf.gather (target dims are dynamic)?

**Comment by drpngx** [comment]
Sorry, what is the problem with `tf.transpose`? Maybe in a larger example?

**Comment by thjashin** [comment]
@drpngx Sorry for not being clear. I mean generating `perm` for tf.transpose() is a little too complex in this simple condition. Could it be possible to add an op that help generate Tensors like `perm`?

**Comment by drpngx** [comment]
@rmlarsen for advice

**Comment by rmlarsen** [comment]
I suppose something like tf.swap(a,i,j), where a, i, j are tensors, that swap a[i, ...] and a[j, ...] would be a reasonable thing to add (sort of a companion to gather).

**Comment by rmlarsen** [comment]
@mrry Maybe Derek has a clever idea how to do this?

**Comment by mrry** [comment]
Nothing springs to mind alas. I did a little search for NumPy implementations of the same operation, and everything I found used in-place updates, which won't really work for the non-Variable-tensor case. I did notice that NumPy has [a `swapaxes()` function](https://docs.scipy.org/doc/numpy-1.10.0/reference/generated/numpy.swapaxes.html), which looks like it would solve the original problem, and if we added one it would keep our API in sync with NumPy.
